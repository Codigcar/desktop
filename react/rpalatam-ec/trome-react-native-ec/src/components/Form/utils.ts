import { useCallback, useState } from 'react'

import { FormErrorsProps, FormErrorsReturn, ValidationErros } from './types'

export const useFormErrors = ({
  ref: formRef,
  schema,
}: FormErrorsProps): FormErrorsReturn => {
  const [validationErrors, setValidationErros] = useState<ValidationErros>({})

  const handleChangeField = useCallback(
    async (name, value) => {
      const error = formRef.current?.getFieldError(name)
      const hasError = error && error.length > 0
      try {
        await schema.validate({ [name]: value })
        if (hasError) {
          formRef.current?.setFieldError(name, '')
          setValidationErros((prev) => ({ ...prev, [name]: '' }))
        }
      } catch (err) {
        if (err.inner[name] && (!error || error !== err.inner[name])) {
          formRef.current?.setFieldError(name, err.inner[name])
          setValidationErros((prev) => ({ ...prev, [name]: err.inner[name] }))
        } else if (!err.inner[name] && hasError) {
          formRef.current?.setFieldError(name, '')
          setValidationErros((prev) => ({ ...prev, [name]: '' }))
        }
      }
    },
    [formRef, schema],
  )

  const setErrors = useCallback(
    (errors) => {
      formRef.current?.setErrors(errors)
      setValidationErros(errors)
    },
    [formRef],
  )

  return {
    errors: validationErrors,
    setFieldError: handleChangeField,
    setErrors,
  }
}
