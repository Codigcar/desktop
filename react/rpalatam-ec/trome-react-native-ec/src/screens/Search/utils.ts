export const computeEstimatedDate = (date: number): string => {
  const currentDate = new Date()
  const approx = currentDate.getDate() + date
  currentDate.setDate(approx)

  const month = `0${currentDate.getUTCMonth() + 1}`.slice(-2)
  const day = `0${currentDate.getUTCDate()}`.slice(-2)
  const year = currentDate.getUTCFullYear()

  return `${year}-${month}-${day}`
}

export const parseDate = (date?: Date): string | null => {
  if (!date) return null
  const month = `0${date.getUTCMonth() + 1}`.slice(-2)
  const day = `0${date.getUTCDate()}`.slice(-2)
  const year = date.getUTCFullYear()

  return `${year}-${month}-${day}`
}

export const timeFrames = [
  {
    value: '',
    label: 'Cualquier fecha',
  },
  {
    value: '-1',
    label: 'Últimas 24 horas',
  },
  {
    value: '-7',
    label: 'Última semana',
  },
  {
    value: '-30',
    label: 'Último mes',
  },
  {
    value: 'custom',
    label: 'Personalizado',
  },
]
