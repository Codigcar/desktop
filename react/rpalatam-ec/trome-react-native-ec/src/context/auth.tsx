import analytics from '@react-native-firebase/analytics'
import crashlytics from '@react-native-firebase/crashlytics'
import { EventEmitter } from 'events'
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState,
} from 'react'
import { Linking } from 'react-native'

import { User } from '../entities/User'
import auth from '../services/auth'
import subscription from '../services/subscription'
import { googleSignOut, init, tokenExchange } from '../utils/auth'
import { ENABLE_AUTHENTICATION } from '../utils/flags'
import { sendPostMessageToWebview } from '../utils/refs'

type AuthListener = (isAuthenticated: boolean) => void

interface Context {
  isAuthenticated: boolean | undefined
  isSubscribed: boolean | undefined
  signin: (provider: string) => void
  signout: () => Promise<void>
  updateUserProfile: (userProfile: User) => Promise<void>
  onAuthenticationChange: (listener: AuthListener) => () => void
  user: User
  verifySubscription: () => Promise<void>
}

const AuthContext = createContext<Context>({} as Context)
const authEmitter = new EventEmitter()
const AUTH_EVENT_NAME = 'auth.event'

export const AuthProvider: React.FC = ({ children }) => {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>()
  const [user, setUser] = useState<User>({})
  const [subscriptor, setSubscriptor] = useState<boolean>()

  const onAuthenticationChange = useCallback((listener: AuthListener) => {
    authEmitter.on(AUTH_EVENT_NAME, listener)
    return function () {
      authEmitter.off(AUTH_EVENT_NAME, listener)
    }
  }, [])

  useEffect(() => {
    init()
    const restoreSession = async () => {
      try {
        if (!ENABLE_AUTHENTICATION) return setIsAuthenticated(false)
        const token = await auth.getToken()
        if (!token) return setIsAuthenticated(false)
        setIsAuthenticated(true)
      } catch (error) {
        if (error instanceof Error) crashlytics().recordError(error)
        setIsAuthenticated(false)
      }
    }
    restoreSession()
  }, [])

  const signin = useCallback((_: string) => {
    setIsAuthenticated(true)
    authEmitter.emit(AUTH_EVENT_NAME, true)
  }, [])

  const signout = useCallback(async () => {
    await auth.logout()
    await googleSignOut()
    await analytics().logEvent('logout')
    setIsAuthenticated(false)
    sendPostMessageToWebview('Home', { type: 'auth.LOGOUT' })
    authEmitter.emit(AUTH_EVENT_NAME, false)
  }, [])

  const updateUserProfile = useCallback(async (data: User) => {
    sendPostMessageToWebview('Home', {
      type: 'auth.PROFILE',
      payload: { ...data },
    })
    setUser(data)
  }, [])

  useEffect(() => {
    if (typeof isAuthenticated === 'undefined') return
    if (isAuthenticated) {
      auth
        .getUserProfile()
        .then((profile) => {
          sendPostMessageToWebview('Home', {
            type: 'auth.PROFILE',
            payload: { ...profile },
          })
          setUser(profile)
        })
        .catch((error) => {
          signout()
          crashlytics().recordError(error)
        })
    } else {
      setUser({})
    }
  }, [isAuthenticated, signout])

  const verifySubscription = useCallback(async () => {
    if (!subscription) return
    try {
      const params: Record<string, string | undefined> = { id: user.id }
      const status = await subscription.checkAccess(params)
      setSubscriptor(status)
    } catch (error) {
      if (error instanceof Error) crashlytics().recordError(error)
    }
  }, [user.id])

  useEffect(() => {
    if (typeof isAuthenticated === 'undefined') return
    if (isAuthenticated && !user.id) return
    if (isAuthenticated && user.id) {
      verifySubscription()
    } else {
      setSubscriptor(false)
    }
  }, [isAuthenticated, user.id, verifySubscription])

  useEffect(() => {
    if (typeof isAuthenticated === 'undefined') return
    if (isAuthenticated && !user.id) return
    const id = user.id || null
    analytics().setUserId(id)
    crashlytics().setUserId(id || '')
  }, [isAuthenticated, user.id])

  useEffect(() => {
    if (typeof subscriptor === 'undefined') return
    const type = subscriptor ? 'premium' : 'free'
    analytics().setUserProperty('account_type', type)
    sendPostMessageToWebview('Home', {
      type: 'subscription.UPDATE_STATUS',
      payload: { status: subscriptor },
    })
  }, [subscriptor])

  useEffect(() => {
    const callback = async (params: { url: string }) => {
      try {
        const config = auth.getOAuthConfig?.()
        if (!config) return
        if (!params.url.startsWith(config.redirect_url)) return
        const token = await tokenExchange(config, params.url)
        await auth.setToken(token)
        signin('oauth')
      } catch (_) {}
    }
    const listener = Linking.addEventListener('url', callback)
    return () => {
      listener.remove()
    }
  }, [signin])

  return (
    <AuthContext.Provider
      value={{
        isAuthenticated,
        isSubscribed: subscriptor,
        signin,
        signout,
        updateUserProfile,
        onAuthenticationChange,
        user,
        verifySubscription,
      }}>
      {children}
    </AuthContext.Provider>
  )
}

export const useAuth = (): Context => useContext(AuthContext)
