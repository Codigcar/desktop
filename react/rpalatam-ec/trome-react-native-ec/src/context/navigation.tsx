import AsyncStorage from '@react-native-async-storage/async-storage'
import analytics from '@react-native-firebase/analytics'
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react'

import * as utils from '../utils/categories'
import { STORE_CATEGORIES } from '../utils/constants'
import { NavigationItem, getMainNavigation } from '../utils/navigation'
import { HomeWebviewRef } from '../utils/refs'
import type { NavCategory } from '../utils/categories'

type SetListOfCategory =
  | NavCategory[]
  | ((prevState: NavCategory[]) => NavCategory[])

type Context = {
  categories: NavCategory[]
  mainNavigation: NavigationItem[]
  sendCategoriesToWebview: () => void
  setListOfCategory: (categories: SetListOfCategory) => Promise<void>
}

const NavigationContext = createContext<Context>({} as Context)

export const NavigationProvider: React.FC = ({ children }) => {
  const [mainNavigation, setMainNavigation] = useState<NavigationItem[]>([])
  const [categories, setCategories] = useState<NavCategory[]>([])
  const refCategories = useRef<NavCategory[]>([])

  const sendCategoriesToWebview = useCallback(() => {
    HomeWebviewRef.current?.injectJavaScript(`(function() {
      if (window.NATIVE_CONNECTION && window.NATIVE_CONNECTION.categories) {
        window.NATIVE_CONNECTION.categories.setCategoriesConfig(${JSON.stringify(
          categories,
        )});
      }
    })();`)
  }, [categories])

  const setListOfCategory = useCallback(
    async (listOfCategory: SetListOfCategory) => {
      const { current } = refCategories
      const isFn = typeof listOfCategory === 'function'
      const list = isFn ? listOfCategory(current) : listOfCategory

      const changed = list.find((category) => {
        const item = current.find(({ key }) => key === category.key)
        if (!item) return false
        return !item.active !== !category.active
      })
      if (changed) {
        const name = changed.active ? 'add_to_home' : 'remove_from_home'
        await analytics().logEvent(name, {
          id: changed.key,
          name: changed.label,
          type: 'section',
        })
      }

      await AsyncStorage.setItem(STORE_CATEGORIES, JSON.stringify(list))
      setCategories(list)
    },
    [],
  )

  const initNavigation = useCallback(async () => {
    const [navigation, localCategories] = await Promise.all([
      getMainNavigation(),
      utils.loadLocalCategories(),
    ])
    setMainNavigation(navigation)

    const categoryList = navigation
      .filter((item) => item.type === 'category')
      .map((item, index) => ({
        active: index < 7,
        required: item.key === 'portada',
        ...item,
      }))

    if (refCategories.current.length === 0 && !localCategories) {
      setCategories(categoryList)
      return
    }

    const composedCategories = utils.composeCategoriesWithLocal(
      categoryList,
      localCategories || refCategories.current,
    )
    setCategories(composedCategories)
  }, [])

  useEffect(() => {
    initNavigation()
  }, [initNavigation])

  useEffect(() => {
    if (categories.length > 0) sendCategoriesToWebview()
    refCategories.current = categories
  }, [categories, sendCategoriesToWebview])

  return (
    <NavigationContext.Provider
      value={{
        categories,
        mainNavigation,
        sendCategoriesToWebview,
        setListOfCategory,
      }}>
      {children}
    </NavigationContext.Provider>
  )
}

export const useMainNavigation = (): Context => useContext(NavigationContext)
