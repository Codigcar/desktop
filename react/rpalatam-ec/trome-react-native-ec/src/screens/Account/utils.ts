import { App } from '../../utils/config'
import * as Schema from '../../utils/validation'

const name = Schema.string()
  .matches(/\d+/i, 'Este campo no puede contener números')
  .matches(/\D{51,}/i, 'Máximo 50 caracteres')

export const ProfileSchema = Schema.object({
  first_name: App.select({
    elcomercio: name.clone().min(2).required(),
    default: name.clone().required(),
  }),
  last_name: name.clone().min(2).required(),
  second_last_name: name.clone().min(2),
  email: Schema.string().required().email(),
  mobile_phone: App.select({
    elcomercio: Schema.string()
      .matches(/\D+/i, 'Se aceptan solo números')
      .min(6, 'Mínimo 6 dígitos')
      .max(12, 'Máximo 12 dígitos'),
    default: Schema.string()
      .min(9, 'Mínimo 9 dígitos')
      .max(13, 'Máximo 13 dígitos'),
  }),
  user_metadata: {
    documentNumber: App.select({
      elcomercio: Schema.string().required(),
      default: Schema.string(),
    })
      .matches(/\s+/i, 'No se permiten espacios en blanco')
      .when('user_metadata.documentType', (value, schema) => {
        if (value === undefined) return schema
        if (value === 'DNI') {
          return schema
            .required()
            .matches(/\D+/i, 'Se aceptan solo números')
            .min(8, 'Se aceptan 8 dígitos')
            .max(8, 'Se aceptan 8 dígitos')
        }
        return schema
          .required()
          .min(9, 'Mínimo 9 dígitos')
          .max(20, 'Máximo 20 dígitos')
      }),
    documentType: Schema.string(),
  },
})
const gendersBase = [
  { label: 'Hombre', value: 'MALE' },
  { label: 'Mujer', value: 'FEMALE' },
]
export const OPTIONS = {
  country: [{ label: 'Perú', value: '260000' }],
  civilStatus: [
    { label: 'Soltero (a)', value: 'SO' },
    { label: 'Casado (a)', value: 'CA' },
    { label: 'Divorciado (a)', value: 'DI' },
    { label: 'Viudo (a)', value: 'VI' },
  ],
  document: [
    { label: 'DNI', value: 'DNI' },
    { label: 'CEX', value: 'CEX' },
    { label: 'CDI', value: 'CDI' },
  ],
  gender: App.select({
    elcomercio: [
      ...gendersBase,
      { label: 'Prefiero no decirlo', value: 'UNKNOWN' },
      { label: 'Otro', value: 'OTHER' },
    ],
    default: gendersBase,
  }),
}
