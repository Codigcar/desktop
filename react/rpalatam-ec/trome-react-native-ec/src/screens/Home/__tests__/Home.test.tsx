import AsyncStorage from '@react-native-async-storage/async-storage'
import { act, render } from '@testing-utils/library'
import React from 'react'
import { Alert, Share } from 'react-native'
import { Interstitial } from 'react-native-ad-manager'

import { useAuth } from '../../../context/auth'
import { useFavorites } from '../../../context/favorites'
import { useMainNavigation } from '../../../context/navigation'
import { useTopics } from '../../../context/topics'
import { getRemoteValue, requestUserPermission } from '../../../utils/firebase'
import { checkNeedsUpdate } from '../../../utils/inAppUpdates'
import { sendFeedbackByEmail } from '../../../utils/mailer'
import {
  HomeWebviewRef,
  PaywallWebviewRef,
  sendPostMessageToWebview,
} from '../../../utils/refs'
import HomeScreen from '../Home'
import HOME_SCREEN_SHOWN from '../homeShown'

jest.mock('@react-native-firebase/crashlytics', () => () => ({
  log: jest.fn(),
  recordError: jest.fn(),
}))

const mockSetMessagesDisplaySuppressed = jest.fn()
const mockTriggerEvent = jest.fn()
jest.mock('@react-native-firebase/in-app-messaging', () => () => ({
  setMessagesDisplaySuppressed: mockSetMessagesDisplaySuppressed,
  triggerEvent: mockTriggerEvent,
}))

const mockNavigate = jest.fn()
const mockDispatch = jest.fn()
jest.mock('@react-navigation/native', () => ({
  DrawerActions: { openDrawer: jest.fn().mockReturnValue('open') },
  useNavigation: jest.fn(() => ({
    addListener: jest.fn(),
    dispatch: mockDispatch,
    removeListener: jest.fn(),
    navigate: mockNavigate,
  })),
}))

jest.mock('react-native-store-review', () => ({
  isAvailable: true,
  requestReview: jest.fn(),
}))
// eslint-disable-next-line @typescript-eslint/no-var-requires
const requestReview = require('react-native-store-review').requestReview

jest.mock('../../../context/auth')
const mockUseAuth = useAuth as jest.Mock

jest.mock('../../../context/favorites')
const mockUseFavorites = useFavorites as jest.Mock

jest.mock('../../../context/navigation')
const mockUseMainNavigation = useMainNavigation as jest.Mock

jest.mock('../../../context/topics')
const mockUseTopics = useTopics as jest.Mock

jest.mock('../../../utils/auth')

jest.mock('../../../utils/firebase')
const mockGetRemoteValue = getRemoteValue as jest.Mock
const mockRequestUserPermission = requestUserPermission as jest.Mock

jest.mock('../../../utils/inAppUpdates')
const mockCheckNeedsUpdate = checkNeedsUpdate as jest.Mock

jest.mock('../../../utils/mailer')

jest.mock('../../../utils/refs', () => ({
  HomeWebviewRef: jest.fn(),
  PaywallWebviewRef: jest.fn(),
  sendPostMessageToWebview: jest.fn(),
}))

jest.useFakeTimers()

beforeAll(() => {
  Object.defineProperty(HomeWebviewRef, 'current', {
    value: { injectJavaScript: jest.fn() },
  })
})

beforeEach(() => {
  jest.spyOn(HOME_SCREEN_SHOWN, 'get').mockReturnValue(true)
  mockGetRemoteValue.mockImplementation(() => ({ asString: jest.fn() }))
  mockRequestUserPermission.mockResolvedValue(() => ({ finally: jest.fn() }))
  mockUseAuth.mockImplementation(() => jest.fn())
  mockUseFavorites.mockImplementation(() => jest.fn())
  mockUseMainNavigation.mockImplementation(() => ({ categories: [] }))
  mockUseTopics.mockReturnValue({ setDefaultTopics: jest.fn() })
})

describe('Home', () => {
  function onMessage<T>(data: T): () => void {
    const utils = render(<HomeScreen />)
    const webview = utils.getByTestId('home-screen')
    const nativeEvent = { data: JSON.stringify(data) }
    return () => webview.props.onMessage({ nativeEvent })
  }

  it('Mount screen', async () => {
    const set = jest.spyOn(HOME_SCREEN_SHOWN, 'set')
    const { getByTestId, toJSON } = render(<HomeScreen />)

    await act(async () => {
      expect(toJSON()).toMatchSnapshot()
    })

    const webview = getByTestId('home-screen')
    expect(webview.props.source.uri).toBe(
      'https://pwa.dev.elcomercio.pe/#?version=50',
    )
    expect(set).toBeCalledWith(true)
  })

  describe('ad interstitial', () => {
    beforeEach(async () => {
      await AsyncStorage.clear()
      mockUseFavorites.mockImplementation(() => ({
        sendFavoritesToWebview: jest.fn(() => ({ ids: jest.fn() })),
      }))
      mockGetRemoteValue.mockImplementation((key) => ({
        asString: jest.fn(() =>
          key === 'interstitial_home' ? 'enabled' : 'disabled',
        ),
      }))
      mockCheckNeedsUpdate.mockResolvedValue(false)
      mockUseAuth.mockImplementation(() => ({
        isSubscribed: false,
      }))
      jest.spyOn(Interstitial, 'requestAd').mockClear()
    })

    it('do request interstitial', async () => {
      const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
      await act(async () => sendMessage())
      expect(Interstitial.requestAd).toBeCalled()
    })

    it('do not request when there is an update', async () => {
      mockCheckNeedsUpdate.mockResolvedValueOnce(true)
      const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
      await act(async () => sendMessage())
      expect(Interstitial.requestAd).not.toBeCalled()
    })

    it('do not request when user is a subscriber', async () => {
      mockUseAuth.mockImplementation(() => ({
        isSubscribed: true,
      }))
      const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
      await act(async () => sendMessage())
      expect(Interstitial.requestAd).not.toBeCalled()
    })

    it('do not request when HOME_SCREEN_SHOWN return false', async () => {
      jest.spyOn(HOME_SCREEN_SHOWN, 'get').mockReturnValue(false)
      const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
      await act(async () => sendMessage())
      expect(Interstitial.requestAd).not.toBeCalled()
    })

    it('do not request when disabled', async () => {
      mockGetRemoteValue.mockImplementation(() => ({
        asString: jest.fn(() => 'disabled'),
      }))
      const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
      await act(async () => sendMessage())
      expect(Interstitial.requestAd).not.toBeCalled()
    })
  })

  describe('handle postmessage', () => {
    describe('app init', () => {
      beforeEach(() => {
        mockUseFavorites.mockImplementation(() => ({
          sendFavoritesToWebview: jest.fn(() => ({ ids: jest.fn() })),
        }))
      })

      it('send categories to webview', async () => {
        const fnSendCategoriesToWebview = jest.fn()
        mockUseMainNavigation.mockImplementationOnce(() => ({
          categories: [{ id: 'id' }],
          sendCategoriesToWebview: fnSendCategoriesToWebview,
        }))
        const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
        await act(async () => sendMessage())
        expect(fnSendCategoriesToWebview).toBeCalled()
      })

      it('send favorites to webview', async () => {
        const fnIds = jest.fn()
        mockUseFavorites.mockImplementationOnce(() => ({
          sendFavoritesToWebview: jest.fn(() => ({ ids: fnIds })),
        }))
        const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
        await act(async () => sendMessage())
        expect(fnIds).toBeCalled()
      })

      it('set user to webview', async () => {
        const user = { id: 'abc' }
        mockUseAuth.mockImplementationOnce(() => ({
          isAuthenticated: true,
          user,
        }))
        const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
        await act(async () => sendMessage())
        expect(HomeWebviewRef.current?.injectJavaScript).toHaveBeenCalledWith(
          expect.stringContaining(
            JSON.stringify({ type: 'auth.PROFILE', payload: user }),
          ),
        )
      })

      it('send topics to webview', async () => {
        const topics = ['id']
        mockUseTopics.mockReturnValue({ topics, setDefaultTopics: jest.fn() })
        const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
        await act(async () => sendMessage())
        expect(HomeWebviewRef.current?.injectJavaScript).toHaveBeenCalledWith(
          expect.stringContaining(`setTopics(${JSON.stringify(topics)})`),
        )
      })
    })

    describe('app addons', () => {
      it('open alert', async () => {
        const spy = jest.spyOn(Alert, 'alert')
        const payload = {
          title: 'Test',
          message: 'message',
          buttons: [{ text: 'Cancel' }, { text: 'Remove' }],
          options: { cancelable: true },
        }
        const sendMessage = onMessage({ type: 'app.SEND_ALERT', payload })
        await act(async () => sendMessage())
        const [title, message, buttons, options] = spy.mock.calls[0]
        expect(title).toBe(payload.title)
        expect(message).toBe(payload.message)

        buttons?.forEach((button) => {
          button.onPress?.()
          expect(sendPostMessageToWebview).toHaveBeenLastCalledWith('Home', {
            payload: { text: button.text },
            type: 'app.ON_PRESS_ALERT',
          })
        })

        expect(options).toEqual(expect.objectContaining(payload.options))
        options?.onDismiss?.()
        expect(sendPostMessageToWebview).toHaveBeenLastCalledWith('Home', {
          payload: payload.options,
          type: 'app.ON_DISMISS_ALERT',
        })
      })

      it('open sheet', async () => {
        const { getByTestId, queryByTestId } = render(<HomeScreen />)
        const webview = getByTestId('home-screen')
        const nativeEvent = {
          data: JSON.stringify({ type: 'app.OPEN_MODE_SHEET' }),
        }
        expect(queryByTestId('backdrop-touchable')).toBeNull()
        await act(async () => webview.props.onMessage({ nativeEvent }))
        act(() => jest.runAllTimers())
        expect(getByTestId('backdrop-touchable')).toBeDefined()
      })

      it('send feedback by email', async () => {
        const user = { id: 'id', email: 'ec@ec.pe' }
        mockUseAuth.mockImplementationOnce(() => ({ user }))
        const sendMessage = onMessage({ type: 'app.SEND_FEEDBACK_BY_EMAIL' })
        await act(async () => sendMessage())
        expect(sendFeedbackByEmail).toBeCalledWith(user)
      })

      it('store review', async () => {
        const sendMessage = onMessage({ type: 'app.STORE_REVIEW' })
        await act(async () => sendMessage())
        expect(requestReview).toBeCalled()
      })
    })

    it('categories', async () => {
      const fnSetListOfCategory = jest.fn()
      mockUseMainNavigation.mockImplementationOnce(() => ({
        setListOfCategory: fnSetListOfCategory,
      }))
      const payload = { categories: [{ id: 'id' }] }
      const sendMessage = onMessage({
        type: 'categories.SET_LIST_OF_CATEGORY',
        payload,
      })
      await act(async () => sendMessage())
      expect(fnSetListOfCategory).toBeCalledWith(payload.categories)
    })

    describe('favorites', () => {
      it('get story', async () => {
        const fnStories = jest.fn()
        const stories = [{ id: 'id' }]
        mockUseFavorites.mockImplementationOnce(() => ({
          setFavorites: jest.fn(() => ({ stories: fnStories })),
          fetchFavoritesStories: jest.fn().mockResolvedValueOnce(stories),
        }))
        const sendMessage = onMessage({ type: 'favorites.GET_STORIES_BY_ID' })
        await act(async () => sendMessage())
        expect(fnStories).toBeCalledWith(stories)
      })

      it('set list of favorites', async () => {
        const fnIds = jest.fn()
        mockUseFavorites.mockImplementationOnce(() => ({
          setFavorites: jest.fn(() => ({ ids: fnIds })),
        }))
        const payload = { ids: ['1', '2'] }
        const sendMessage = onMessage({
          type: 'favorites.SET_LIST_OF_FAVORITES',
          payload,
        })
        await act(async () => sendMessage())
        expect(fnIds).toBeCalledWith(payload.ids)
      })
    })

    describe('in app messaging', () => {
      it('hide messages', async () => {
        const payload = { enabled: true }
        const sendMessage = onMessage({
          type: 'inAppMessaging.HIDE_MESSAGES',
          payload,
        })
        await act(async () => sendMessage())
        expect(mockSetMessagesDisplaySuppressed).toHaveBeenLastCalledWith(
          payload.enabled,
        )
      })

      it('trigger event', async () => {
        const payload = { eventId: 'id' }
        const sendMessage = onMessage({
          type: 'inAppMessaging.TRIGGER_EVENT',
          payload,
        })
        await act(async () => sendMessage())
        expect(mockTriggerEvent).toHaveBeenLastCalledWith(payload.eventId)
      })
    })

    describe('navigation', () => {
      it('navigate to screen', async () => {
        const payload = { name: 'Stack', params: { screen: 'Screen' } }
        const sendMessage = onMessage({
          type: 'navigation.NAVIGATE_TO_SCREEN',
          payload,
        })
        await act(async () => sendMessage())
        expect(mockNavigate).toHaveBeenLastCalledWith(
          payload.name,
          payload.params,
        )
      })

      it('open drawer', async () => {
        const sendMessage = onMessage({ type: 'navigation.OPEN_DRAWER' })
        await act(async () => sendMessage())
        expect(mockDispatch).toHaveBeenLastCalledWith('open')
      })
    })

    it('paywall screen', async () => {
      const fnInjectJavacript = jest.fn()
      Object.defineProperty(PaywallWebviewRef, 'current', {
        value: { injectJavaScript: fnInjectJavacript },
      })
      const payload = { message: 'message' }
      const sendMessage = onMessage({
        type: 'paywall.SEND_POSTMESSAGE_TO_PAYWALL_SCREEN',
        payload,
      })
      await act(async () => sendMessage())
      expect(fnInjectJavacript).toHaveBeenLastCalledWith(
        expect.stringContaining(JSON.stringify(payload)),
      )
    })

    it('share', async () => {
      const spy = jest.spyOn(Share, 'share')
      const content = { title: 'title', url: 'https://ec.pe/' }
      const sendMessage = onMessage({ type: 'onShare', content })
      await act(async () => sendMessage())
      expect(spy).toBeCalledWith({ message: `${content.title} ${content.url}` })
    })

    describe('subscription', () => {
      it('check access for user who is anonymous', async () => {
        mockUseAuth.mockImplementationOnce(() => ({
          isSubscribed: false,
          user: {},
        }))
        const sendMessage = onMessage({ type: 'subscription.CHECK_ACCESS' })
        await act(async () => sendMessage())
        expect(mockNavigate).toHaveBeenLastCalledWith('Auth', {
          screen: 'InitialScreen',
        })
        expect(
          HomeWebviewRef.current?.injectJavaScript,
        ).toHaveBeenLastCalledWith(
          expect.stringContaining(
            JSON.stringify({
              type: 'subscription.CHECK_ACCESS_CALLBACK',
              payload: { status: false },
            }),
          ),
        )
      })

      it('check access for user who is subscriber', async () => {
        const fnVerifySubscription = jest.fn().mockResolvedValue(null)
        mockUseAuth.mockImplementationOnce(() => ({
          isSubscribed: true,
          user: { id: '123' },
          verifySubscription: fnVerifySubscription,
        }))

        const sendMessage = onMessage({ type: 'subscription.CHECK_ACCESS' })
        await act(async () => sendMessage())
        expect(fnVerifySubscription).toBeCalled()
        expect(
          HomeWebviewRef.current?.injectJavaScript,
        ).toHaveBeenLastCalledWith(
          expect.stringContaining(
            JSON.stringify({
              type: 'subscription.CHECK_ACCESS_CALLBACK',
              payload: { status: true },
            }),
          ),
        )
      })
    })

    describe('topics', () => {
      it('set default topics', async () => {
        mockUseFavorites.mockReturnValueOnce({
          sendFavoritesToWebview: jest.fn(() => ({ ids: jest.fn() })),
        })
        mockRequestUserPermission.mockResolvedValue(true)
        const setDefaultTopics = jest.fn()
        mockUseTopics.mockReturnValue({
          setDefaultTopics,
          setTopics: jest.fn(),
        })
        const sendMessage = onMessage({ type: 'init.WEB_LOADED' })
        await act(async () => sendMessage())
        expect(setDefaultTopics).toBeCalled()
      })

      it('subscribe to topic', () => {
        const subscribeToTopic = jest.fn()
        mockUseTopics.mockReturnValueOnce({ subscribeToTopic })
        onMessage({
          type: 'topics.SUBSCRIBE_TO_TOPIC',
          payload: { topic: 'topic' },
        })()
        expect(subscribeToTopic).toBeCalledWith('topic')
      })

      it('unsubscribe from topic', () => {
        const unsubscribeFromTopic = jest.fn()
        mockUseTopics.mockReturnValueOnce({ unsubscribeFromTopic })
        onMessage({
          type: 'topics.UNSUBSCRIBE_FROM_TOPIC',
          payload: { topic: 'topic' },
        })()
        expect(unsubscribeFromTopic).toBeCalledWith('topic')
      })
    })
  })
})
