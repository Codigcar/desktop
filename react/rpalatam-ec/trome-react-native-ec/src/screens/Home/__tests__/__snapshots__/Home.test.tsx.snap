// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Mount screen 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {},
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFCB05",
          "display": "flex",
          "flex": 1,
          "overflow": "hidden",
        },
      ]
    }
  >
    <RNCSafeAreaView
      edges={
        Array [
          "top",
        ]
      }
      style={
        Array [
          Object {
            "backgroundColor": "#FFCB05",
            "display": "flex",
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <RNCWebView
        allowsFullscreenVideo={true}
        allowsInlineMediaPlayback={true}
        cacheEnabled={true}
        injectedJavaScriptBeforeContentLoaded="
  window.__REACT_NATIVE_CONTEXT_TOPICS__ = true;
  window.__REACT_NATIVE_SCREENS_AUTH_SIGNWALL__ = true;
  window.__REACT_NATIVE_SCREENS_PAYWALL_MODAL__ = true;
  window.__REACT_NATIVE_SCREENS_STORY__ = true;
  true;
"
        injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
        injectedJavaScriptForMainFrameOnly={true}
        javaScriptEnabled={true}
        mediaPlaybackRequiresUserAction={true}
        messagingEnabled={true}
        onContentProcessDidTerminate={[Function]}
        onHttpError={[Function]}
        onLoadingError={[Function]}
        onLoadingFinish={[Function]}
        onLoadingProgress={[Function]}
        onLoadingStart={[Function]}
        onMessage={[Function]}
        onShouldStartLoadWithRequest={[Function]}
        source={
          Object {
            "uri": "https://pwa.dev.elcomercio.pe/#?version=50",
          }
        }
        style={
          Array [
            Object {
              "flex": 1,
              "overflow": "hidden",
            },
            Object {
              "backgroundColor": "#ffffff",
            },
            Object {
              "flex": 1,
              "opacity": 0.99,
            },
          ]
        }
        testID="home-screen"
        textInteractionEnabled={true}
        useSharedProcessPool={true}
      />
    </View>
  </View>
</RNCSafeAreaProvider>
`;
